chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -2127,14 +2133,92 @@ impl PeerManagerActor {
-        if self.state.accounts_data.set_keys(info.0.tier1_accounts.clone()) {
-            // TODO(gprusak): the set of tier1 accounts has changed.
-            // We might miss some data, so start a full sync with connected peers.
-        }
-        *self.state.chain_info.write() = info.0;
+        let now = self.clock.now_utc();
+        let info = info.0;
+        let state = self.state.clone();
+        let event_sink = self.event_sink.clone();
+        tokio::spawn(async move {
+            *state.chain_info.write() = info.clone();
+            // If the key set didn't change, early exit.
+            if !state.accounts_data.set_keys(info.tier1_accounts.clone()) {
+                return;
+            }
+            // If the set of keys has changed, and the node is a validator,
+            // we should try to sign data and broadcast it. However, this is
+            // also a trigger for a full sync, so a dedicated broadcast is
+            // not required.
+            //
+            // TODO(gprusak): For dynamic self-IP-discovery, add a STUN daemon which
+            // will add new AccountData and trigger an incremental broadcast.
+            if let Some(vc) = &state.config.validator {
+                let my_account_id = vc.signer.validator_id();
+                let my_public_key = vc.signer.public_key();
+                // TODO(gprusak): STUN servers should be queried periocally by a daemon
+                // so that the my_peers list is always resolved.
+                // Note that currently we will broadcast an empty list.
+                // It won't help us to connect the the validator BUT it
+                // will indicate that a validator is misconfigured, which
+                // is could be useful for debugging. Consider keeping this
+                // behavior for situations when the IPs are not known.
+                let my_peers = match &vc.endpoints {
+                    config::ValidatorEndpoints::TrustedStunServers(_) => vec![],
+                    config::ValidatorEndpoints::PublicAddrs(addrs) => {
+                        addrs.iter().cloned().map(|addr| PeerAddr { addr, peer_id: None }).collect()
+                    }
+                };
+                let my_data = info.tier1_accounts.iter().filter_map(|((epoch_id,account_id),key)| {
+                    if account_id != my_account_id{
+                        return None;
+                    }
+                    if key != &my_public_key {
+                        warn!(target: "network", "node's account_id found in TIER1 accounts, but the public keys do not match");
+                        return None;
+                    }
+                    Some(AccountData {
+                        epoch_id: epoch_id.clone(),
+                        account_id: my_account_id.clone(),
+                        timestamp: now,
+                        peers: my_peers.clone(),
+                    }.sign(vc.signer.as_ref()).unwrap())
```

Comment:
This unwrap is suspicious. I think nothing prevents `vc.endpoints` to contain a *reealy*  long list of endpoints?  If unwrap here, let's maybe, just in case, take only the first 10 ips or something, and emit an `traicing::error` otherwise?

Reply:
This would be misconfiguration and should cause a shutdown. However, I should probably move it to config parsing though.

Reply:
I've moved the validation to JSONConfig -> Config convertion (which should be rather read as RawConfig -> ValidatedConfig conversion, probably would be nice to rename these). Both types representing configs are part of near_network package now. My long term plan is to merge near_network_primitives into near_network.